### PR TITLE
Add more detailed notes and status to refunded order with charge authorization

### LINKED ISSUE #757: Stripe can't capture charge for authorized orders after partial refund
#### Affected ticket(s)

1561572-zen
3411693-zen
3146888-zen
7217834-hc
19466492-hc

#### What I expected

When store owner receives On-hold order made via authorized (not captured) Stripe payment, they can refund it partially and then capture the rest amount.

#### What happened instead

When order is On-hold, you can issue a partial refund, but then when changing order status to Processing to capture payment, Stripe is unable to charge customer anymore.

#### Steps to reproduce the issue

Stripe settings: 

Screenshot
Screenshot link: https://d.pr/ZXomSD

1) Place a test order paid with Stripe, it's On-hold now.
2) Refund the order partially via Stripe:

Screenshot
Screenshot link: https://d.pr/kTKlKB

3) Change order status from On-hold to Processing. You'll see the error in notes:

Screenshot
Screenshot link: https://d.pr/3iCPOd

I understand that the _correct_ way of doing this would be to edit On-hold order like this:

Screenshot
Screenshot link: https://d.pr/Xbl8U4

However, from the Edit Order page it's not explicitly clear that by issuing a partial refund, they won't be able to capture a charge anymore.

Setting the priority to high because with this case store owner will need users to pay again.

-------------------

- [ ] Issue assigned to next milestone.
- [ ] Issue assigned a priority (will be assessed by maintainers).

### LINKED ISSUE #1470: Clearer UX for Partial Refunds when funds are still being authorized
**Describe the bug**
When Stripe is set to not capture funds immediately an on-hold order can issue a partial refund and be left with an unclear message that no funds have been exchanged.

**To Reproduce**
1. In Settings uncheck the capture immediately checkbox
2. Create a new order with multiple products
3. Ensure the order status is set to on-hold
4. Issue a partial refund
5. See the message in the order notes that says "Pre-Authorization Released"

**Additional context**
Closely related to #757

### LINKED ISSUE #1480: Add more detailed notes and status to refunded order with charge authorization
When a refund is attempted for an order with a charge that was authorized but not captured:
- change order status to cancelled
- prevent Woo from processing the refund, so that it doesn't show up as refunded on the order
- update order notes as shown in this screenshot:

Screen Shot 2021-02-17 at 8 45 37 PM

Steps to repro to see the above:
1. Change Stripe settings to not capture charges right away
2. Make a test order as a customer with two items
3. As store owner, refund one of the items
4. See order note "Pre-Authorization Released"
5. When webhook comes though, see order note "Pre-Authorization Released via Stripe Dashboard"

### PR DESCRIPTION
**Pull request by ricardo**
# Changes proposed in this Pull Request:

Fixes #1480
Fixes #1470
Fixes #757

- Add more detailed notes to refunded orders with charge authorization.
- Automatically cancel order when attempting to refund an order with charge authorization.
- Prevent Woo from creating a refund for an order with charge authorization.

This part of the code was kinda messy to work with. I tried to improve readability by adding some comments. Please let me know if you have questions about the changes in the code.

# Testing instructions

- Disable "Capture charge immediately" from the plugin settings.
- Enable test mode, add test secret keys and set up a test webhook secret key.
- Make a test order with two items.
- As store owner, refund one of the items. Notice the following result:
  - The order must get cancelled.
  - A note is added: `Pre-Authorization for [value] voided.`.
  - The refund is not added in order details.
- Make a new order and cancel it from the Stripe dashboard. Notice the following result:
  - The order must get cancelled.
  - A different note is added: `Pre-Authorization for [value] voided from the Stripe Dashboard.`.
  - The refund is not added in order details.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).

### CODE REVIEW

**Comment by v18** [comment]
I think this PR will also solve #757

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php
Comment (diegocurbelo) [inline includes/abstracts/abstract-wc-stripe-payment-gateway.php]: This should be in the same line, but we can wait until we merge #1445 and fix it then

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php
Comment (diegocurbelo) [inline includes/abstracts/abstract-wc-stripe-payment-gateway.php]: Maybe replace `dollar amount` with `1) amount (including currency symbol)` as the amount might be in another currency

**Review by diegocurbelo — APPROVED** [review]
I tested the 2 flows and both work as expected.

The code looks good to me 🚢 .

@ricardo, we need to add changelog and readme entries, and then we can merge it.

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php
Comment (ricardo) [reply includes/abstracts/abstract-wc-stripe-payment-gateway.php]: 😀  Will change it.

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php
Comment (ricardo) [reply includes/abstracts/abstract-wc-stripe-payment-gateway.php]: I agree your suggestion sounds better. I copied `dollar amount` from another translation in the same file. I'm gonna update that one too.

**Review by diegocurbelo — APPROVED** [review]

**Pull request merged**
